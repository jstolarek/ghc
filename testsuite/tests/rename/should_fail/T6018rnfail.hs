{-# LANGUAGE TypeFamilies #-}

module T6018rnfail where

-- ID = injectivity declaration `| foo -> bar`

-- use incorrect tyvar in LHS of ID
type family F a = r | a -> a
type family Fc a = r | a -> a where
  Fc a = a
class Fcl a where
  type Ft a = r | a -> a

-- declare result tyvar to be duplicate (without ID)
type family G a = a
type family Gc a = a where
  Gc a = a
class Gcl a where
  type Gt a = a

-- declare result tyvar to be duplicate (with ID)
type family Gb a = a | a -> a
type family Gcb a = a | a -> a where
  Gcb a = a
class Gclb a where
  type Gtb a = a | a -> a

-- incorrect order of tyvars in RHS of ID
type family H a b = r | r -> b a
type family Hc a b = r | r -> b a where
  Hc a b = a
class Hcl a b where
  type Ht a b = r | r -> b a

-- not in-scope tyvar in RHS of ID
type family I a b = r | r -> c
type family Ic a b = r | r -> c where
  Ic a b = a
class Icl a b where
  type It a b = r | r -> c

-- repeated tyvar on RHS of ID
type family J a b = r | r -> a a
type family Jc a b = r | r -> a a where
  Jc a b = a
class Jcl a b where
  type Jt a b = r | r -> a a

-- too many tyvars on RHS of ID
type family K a b = r | r -> a b b
type family Kc a b = r | r -> a b b where
  Kc a b = a
class Kcl a b where
  type Kt a b = r | r -> a b b

-- not in-scope tyvar in LHS of ID
type family L a b = r | c -> a
type family Lc a b = r | c -> a where
  Lc a b = a
class Lcl a b where
  type Lt a b = r | c -> a
