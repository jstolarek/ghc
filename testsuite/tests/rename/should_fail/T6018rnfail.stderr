
T6018rnfail.hs:8:23:
    Incorrect type variable on the LHS of injectivity condition
         Expected : r
         Actual   : a

T6018rnfail.hs:9:24:
    Incorrect type variable on the LHS of injectivity condition
         Expected : r
         Actual   : a

T6018rnfail.hs:12:19:
    Incorrect type variable on the LHS of injectivity condition
         Expected : r
         Actual   : a

T6018rnfail.hs:15:19:
    Type variable `a' naming a type family result also names one of the arguments.

T6018rnfail.hs:16:20:
    Type variable `a' naming a type family result also names one of the arguments.

T6018rnfail.hs:19:15:
    Type variable `a' naming a type family result also names one of the arguments.

T6018rnfail.hs:22:20:
    Type variable `a' naming a type family result also names one of the arguments.

T6018rnfail.hs:23:21:
    Type variable `a' naming a type family result also names one of the arguments.

T6018rnfail.hs:26:16:
    Type variable `a' naming a type family result also names one of the arguments.

T6018rnfail.hs:29:30:
    Unexpected type variable on the RHS of injectivity condition: `b'
    All variables should be bound in type family head and appear at most once in exactly the same order as they were bound.

T6018rnfail.hs:30:31:
    Unexpected type variable on the RHS of injectivity condition: `b'
    All variables should be bound in type family head and appear at most once in exactly the same order as they were bound.

T6018rnfail.hs:33:26:
    Unexpected type variable on the RHS of injectivity condition: `b'
    All variables should be bound in type family head and appear at most once in exactly the same order as they were bound.

T6018rnfail.hs:36:1: Not in scope: type variable `c'

T6018rnfail.hs:37:1: Not in scope: type variable `c'

T6018rnfail.hs:40:3: Not in scope: type variable `c'

T6018rnfail.hs:43:32:
    Unexpected type variable on the RHS of injectivity condition: `a'
    All variables should be bound in type family head and appear at most once in exactly the same order as they were bound.

T6018rnfail.hs:44:33:
    Unexpected type variable on the RHS of injectivity condition: `a'
    All variables should be bound in type family head and appear at most once in exactly the same order as they were bound.

T6018rnfail.hs:47:28:
    Unexpected type variable on the RHS of injectivity condition: `a'
    All variables should be bound in type family head and appear at most once in exactly the same order as they were bound.

T6018rnfail.hs:50:34:
    Too many type variables on RHS of injectivity condition.
         You listed three : a, b, b
         But at most two are allowed : a, b

T6018rnfail.hs:51:35:
    Too many type variables on RHS of injectivity condition.
         You listed three : a, b, b
         But at most two are allowed : a, b

T6018rnfail.hs:54:30:
    Too many type variables on RHS of injectivity condition.
         You listed three : a, b, b
         But at most two are allowed : a, b

T6018rnfail.hs:57:1: Not in scope: type variable `c'

T6018rnfail.hs:58:1: Not in scope: type variable `c'

T6018rnfail.hs:61:3: Not in scope: type variable `c'
